haystack/document_stores/__init__.py

```diff
@@ -7,7 +10,21 @@
 from haystack.document_stores.es_converter import elasticsearch_index_to_document_store
 from haystack.document_stores.es_converter import open_search_index_to_document_store
 
-from haystack.document_stores.elasticsearch import ElasticsearchDocumentStore
+try:
+    with LazyImport() as elasticsearch_import:
+        # Use appropriate ElasticsearchDocumentStore depending on ES client version
+        from elasticsearch import __version__ as ES_VERSION
+
+        if ES_VERSION[0] == 7:
+            from haystack.document_stores.elasticsearch7 import ElasticsearchDocumentStore  # type: ignore  # pylint: disable=reimported,ungrouped-imports
+        elif ES_VERSION[0] == 8:
+            if not typing.TYPE_CHECKING:
+                from haystack.document_stores.elasticsearch8 import ElasticsearchDocumentStore  # type: ignore  # pylint: disable=reimported,ungrouped-imports
+    elasticsearch_import.check()
+except (ModuleNotFoundError, ImportError):
+    # Use Elasticsearch 7 as default
+    from haystack.document_stores.elasticsearch7 import ElasticsearchDocumentStore
+
```

Comment:
Just a comment on LazyImport usage :slightly_smiling_face: 

LazyImport is little more than a "glorified try-catch", and it's useful when you need to _defer_ the import. In this case, though, the import is needed immediately, and not only so, but it's even try-catched already. So LazyImports provides you no benefit in choosing whether to use ES7 or ES8.

So the statement here should look more like:

```python
try:
    # Use appropriate ElasticsearchDocumentStore depending on ES client version
    from elasticsearch import __version__ as ES_VERSION

    if ES_VERSION[0] == 7:
        from haystack.document_stores.elasticsearch7 import ElasticsearchDocumentStore  # type: ignore  # pylint: disable=reimported,ungrouped-imports
    elif ES_VERSION[0] == 8:
        from haystack.document_stores.elasticsearch8 import ElasticsearchDocumentStore  # type: ignore  # pylint: disable=reimported,ungrouped-imports
    else:
        # Use Elasticsearch 7 as default
        from haystack.document_stores.elasticsearch7 import ElasticsearchDocumentStore
except (ModuleNotFoundError, ImportError):
    # No need to import anything if ES could not be imported
    pass
```

You can then put the LazyImport blocks around ES imports in the two subclasses to catch all direct imports, like `from haystack.document_stores.elasticsearch import ElasticSearchDocumentStore`, which would not be catched by this code.

Hope this helps!

Reply:
Thanks @ZanSara!